File: apps/storefront/lib/ticket-assignment.ts
/**
 * The buyer's Ticket Assignments, as rules rather than as markup (#324, parent
 * #322).
 *
 * WHAT THIS FEATURE IS FOR, ON THIS SURFACE ONLY. A buyer of four tickets has
 * four Tickets that are indistinguishable from one another: the Answer Links
 * deliberately disclose nothing, so "ticket 2 of 4" was the whole of what could
 * be said to tell them apart. Naming an address for each one is what finally
 * lets the buyer say which is which — the address is the label, and showing it
 * back is half the point of the write.
 *
 * IT LIVES BESIDE lib/buyer-answers.ts AND NOT INSIDE IT, because the two are
 * separate features on one payload and are behind SEPARATE FLAGS. A deployment
 * may have Ticket Questions and no assignment, and the page must look exactly as
 * it did before this shipped in that case — so every rule about "is there
 * anything to draw" is asked of assignment on its own terms.
 *
 * THE FLAG IS READ OFF THE PAYLOAD AND NOWHERE ELSE. With
 * TICKET_ASSIGNMENT_ENABLED closed the API omits every assignment field
 * (`omitempty`, on purpose — see service/buyer_answers.go), so an absent
 * `assignment_state` is how this app learns the feature is dark. There is no
 * second copy of the flag in this app, exactly as there is none for Ticket
 * Questions.
 *
 * Pure and dependency-free — no React, no i18n runtime, no fetch — so it is
 * directly unit-testable, exactly as its two siblings are. It returns SHAPES and
 * TOKENS and never a sentence: the Storefront owns the words, in the reader's
 * language, and the API deliberately sends refusal tokens rather than prose.
 */

import type { BuyerTicket } from "@/lib/buyer-answers";

/**
 * The three states of one Ticket, as the API spells them.
 *
 * `accepted` IS UNREACHABLE TODAY and modelled anyway. #324 sends no mail, so
 * there is no Assignment Link to click; #325 is what makes it happen. Drawing it
 * now means the surface that has to show it later is already written, rather
 * than being found and changed under time pressure.
 */
export type AssignmentState = "unassigned" | "assigned" | "accepted";

/**
 * Why a Ticket may not be assigned right now — the API's own tokens
 * (service/ticket_assignment.go), never re-derived here.
 *
 * A CLOSED WINDOW IS NOT A HIDDEN SECTION. A door sale, a reversed purchase and
 * an Event that has already happened all keep showing who was assigned to what;
 * what closes is the input, not the record. The same posture the Answer window
 * takes one file over.
 */
export type AssignmentRefusal = "channel_unsupported" | "sale_reversed" | "event_started";

/**
 * The body of an assignment write. One field, because the buyer names an ADDRESS
 * and nothing else — a name they typed would be a fact about a person recorded
 * from somebody else's memory, and it would go straight into the Organization's
 * guest list.
 */
export type AssignmentBody = { holder_email: string };

/**
 * The longest a Holder address may be (RFC 5321's forward path), matching
 * catalog.MaxHolderEmailLength.
 *
 * Mirrored rather than left to the API because this is the one text field on
 * this platform typed by one person ABOUT A DIFFERENT PERSON: nobody signs in
 * with it and nobody corrects it but the buyer, so a bound at the moment of
 * typing is the only bound there is.
 */
export const MAX_HOLDER_EMAIL_LENGTH = 254;

/**
 * Whether this deployment offers assignment at all for this Ticket.
 *
 * THE ABSENCE OF THE FIELD IS THE ANSWER. With the flag closed the row carries
 * no `assignment_state`, and a page that drew an "add an address" box anyway
 * would be offering a control whose every press is answered 404. Undefined and
 * not empty-string: `unassigned` is a real state that must be drawn.
 */
export function assignmentOffered(ticket: BuyerTicket): boolean {
  return ticket.assignment_state !== undefined;
}

/**
 * Whether any Ticket on this sale offers assignment — i.e. whether this section
 * has an assignment half at all.
 */
export function saleOffersAssignment(tickets: BuyerTicket[]): boolean {
  return tickets.some(assignmentOffered);
}

/**
 * This Ticket's state, or null while the feature is dark.
 *
 * Read off the payload and never derived from the address and the timestamps
 * here. The platform has ONE definition of the state (catalog.AssignmentState)
 * and a browser recomputing it would be a second opinion that first disagrees on
 * the case nobody thinks about — an accepted Ticket whose Holder address is
 * being changed.
 */
export function assignmentStateOf(ticket: BuyerTicket): AssignmentState | null {
  const state = ticket.assignment_state;
  return state === "unassigned" || state === "assigned" || state === "accepted" ? state : null;
}

/**
 * Whether this Ticket is the buyer's own (ADR 0048): its Holder is the buyer.
 * The page says "your ticket" on it, asks the buyer to answer it, and offers
 * no link for it — there is nobody to forward one to.
 */
export function isOwnTicket(ticket: BuyerTicket): boolean {
  return ticket.self_held === true;
}

/** The address the buyer gave this Ticket, or "" while it has none. */
export function holderEmailOf(ticket: BuyerTicket): string {
  return ticket.holder_email ?? "";
}

/** Whether the buyer may type an address for this Ticket right now. */
export function canAssign(ticket: BuyerTicket): boolean {
  return assignmentOffered(ticket) && ticket.assignable === true;
}

/**
 * Why the input is closed, or null when it is open (or when there is no feature
 * here to be closed).
 *
 * An unrecognised token answers null rather than a guess, for ADR 0023's reason
 * one level over: a token this app has never heard of must degrade to saying
 * nothing, not to saying the wrong thing about somebody's purchase.
 */
export function assignmentRefusalOf(ticket: BuyerTicket): AssignmentRefusal | null {
  if (!assignmentOffered(ticket) || ticket.assignable === true) return null;
  const refusal = ticket.assignable_refusal;
  if (
    refusal === "channel_unsupported" ||
    refusal === "sale_reversed" ||
    refusal === "event_started"
  ) {
    return refusal;
  }
  return null;
}

/**
 * The address as it will be stored: trimmed and lowercased, matching
 * platform.NormalizeEmail, which is the single point at which an address is
 * normalised on this platform.
 *
 * IT MATTERS MORE HERE THAN ANYWHERE. #325 mints or matches a Customer from a
 * click at this address; one stored as the buyer capitalised it would fail to
 * match the record that person already has, quietly turning one person into two.
 * Normalising on this side as well is what makes "is this already the address?"
 * below answer the same as the API's own no-op check.
 */
export function normalizeHolderEmail(raw: string): string {
  return raw.trim().toLowerCase();
}

/**
 * The pre-flight verdict on what the buyer typed: an API error code, or null to
 * send it.
 *
 * IT ANSWERS WITH THE API'S OWN CODE and never a sentence of its own — ADR 0023,
 * and the same arrangement lib/tax-id.ts and lib/phone.ts have. One code, one
 * catalog entry, one sentence: a typo caught here reads exactly as a typo caught
 * by the API reads, in whichever language the page is being read in.
 *
 * IT IS DELIBERATELY WEAKER THAN THE API'S CHECK and must stay that way. An
 * address is only really validated by mail arriving at it, and a mirror that
 * refused more than the API does would block a buyer from saving an address the
 * platform would have accepted — a form refusing a correct answer, with no
 * appeal. So it catches only what is certainly not an address: nothing at all,
 * something longer than an address may be, anything with whitespace in it
 * (which is the display-name form `Ana <ana@example.com>` that mail.ParseAddress
 * accepts and catalog.ParseHolderEmail then refuses), and the missing `@`.
 *
 * THE MISSING `@` IS THE ONE WORTH CATCHING. Until #325 mails the address,
 * nobody discovers it is wrong: the buyer sees their own typo echoed back and
 * believes the job done.
 */
export function holderEmailRefusal(raw: string): "INVALID_HOLDER_EMAIL" | null {
  const email = normalizeHolderEmail(raw);
  if (email === "" || email.length > MAX_HOLDER_EMAIL_LENGTH) return "INVALID_HOLDER_EMAIL";
  if (/\s/.test(email)) return "INVALID_HOLDER_EMAIL";
  // Exactly one `@`, with something on each side of it. Nothing is said about
  // what that something may contain — the API decides, and this must not become
  // a second opinion about apostrophes, plus addressing or new top-level
  // domains.
  const parts = email.split("@");
  if (parts.length !== 2 || parts[0] === "" || parts[1] === "") return "INVALID_HOLDER_EMAIL";
  return null;
}

/**
 * Whether sending this would change anything.
 *
 * The API treats re-sending the address a Ticket already carries as a no-op that
 * moves no timestamp, so this is not a correctness guard — it is here so the
 * form can say "that is already this ticket's address" instead of reporting a
 * save that saved nothing. On a REASSIGNMENT the distinction has teeth: a
 * genuine change clears that Ticket's Answers, and a buyer who pressed save on
 * an unchanged field must not be told their answers went with it.
 */
export function isUnchangedAssignment(ticket: BuyerTicket, raw: string): boolean {
  const current = holderEmailOf(ticket);
  return current !== "" && normalizeHolderEmail(raw) === current;
}

/**
 * The body to send, or null when there is nothing worth sending.
 *
 * The address goes NORMALISED rather than as typed. The API normalises it again
 * — it must, since nothing about a browser is trusted — and the two agreeing is
 * what keeps the value echoed back on the next render identical to the value the
 * buyer is looking at in the field.
 */
export function assignmentBodyFor(raw: string): AssignmentBody | null {
  if (holderEmailRefusal(raw) !== null) return null;
  return { holder_email: normalizeHolderEmail(raw) };
}

/**
 * How many of this sale's Tickets carry an address, out of how many could.
 *
 * PARTIAL ASSIGNMENT IS THE NORMAL CASE and not a half-finished one: a sale of
 * four where the buyer knows two addresses is complete as far as they are
 * concerned, and nothing on this page may read as an error because of it. So the
 * summary counts rather than warns, and the denominator is the Tickets that CAN
 * be assigned — counting a door sale's Tickets in a total the buyer can never
 * move would be a page setting somebody a goal it will not let them reach.
 */
export function assignmentTally(tickets: BuyerTicket[]): { assigned: number; total: number } {
  const offered = tickets.filter(assignmentOffered);
  return {
    assigned: offered.filter((ticket) => holderEmailOf(ticket) !== "").length,
    total: offered.length,
  };
}

File: backend/internal/catalog/ticket_assignment.go
package catalog

import (
	"net/mail"
	"time"

	"github.com/peter/ticket_pos/backend/internal/platform"
)

// This file holds everything a Ticket Assignment knows about itself: what a
// Holder address may be, which of three states a Ticket is in, and when a buyer
// may assign at all (#324, parent #322).
//
// ASSIGNMENT IS A FACT ABOUT A TICKET, exactly as an Answer is. The buyer names
// an email address for ONE Ticket — not for the Ticket Sale, which stays whole
// with them, and not for the Payment, which never sees a third party's address
// at all. A buyer of four tickets names four addresses, or two, or none.
//
// A HOLDER IS NOT AN OWNER. Nothing here moves a Ticket Sale, a Sale
// Confirmation, the money or the Reversal Window: the buyer may reassign any
// Ticket at any time, and does not stop being the party of record when they do.
// This is assignment, never transfer — see docs/prd-ticket-assignment.md.

// MaxHolderEmailLength caps a Holder address at the longest an address may be
// (RFC 5321's 254 for a forward path).
//
// It exists because this is the ONE text field on this platform typed by one
// person about a DIFFERENT person, so nothing downstream can sanity-check it
// against anything: nobody signs in with it, nobody corrects it but the buyer,
// and until #325 mails it nobody discovers it is nonsense. A cap is the only
// bound available at the moment of typing.
const MaxHolderEmailLength = 254

// TicketAssignmentState is which of the three states one Ticket is in.
//
// A STRING RATHER THAN AN INT, unlike AnswerRefusal beside it, because this one
// travels on the wire. It is the buyer's answer to "what happened to the address
// I typed", and a page that has to map 0/1/2 onto three words is a page holding
// a second copy of this enum.
type TicketAssignmentState string

const (
	// TicketUnassigned: no address has been given. Today's behaviour, unchanged
	// — the buyer, Event Staff and anyone holding the Answer Link may answer,
	// and this is what most Tickets will be for a long time.
	TicketUnassigned TicketAssignmentState = "unassigned"
	// TicketAssigned: an address has been given and nobody has accepted. In
	// #324 this is the terminal state, because no mail is sent: the address sits
	// on the Ticket, is shown back to the buyer so they can tell their Tickets
	// apart, and reaches nobody. #325 is what mails it.
	TicketAssigned TicketAssignmentState = "assigned"
	// TicketAccepted: a Holder proved the address and is a Customer.
	//
	// UNREACHABLE IN #324 AND DELIBERATELY MODELLED ANYWAY. Nothing in this
	// ticket writes accepted_at — there is no Assignment mail, so there is no
	// link to click. It exists here so that every reader added by this ticket
	// already knows the full state machine, rather than being found and changed
	// when #325 lands the accept flow.
	TicketAccepted TicketAssignmentState = "accepted"
)

// AssignmentState reads a Ticket's state off the two timestamps and the address.
//
// DERIVED AND NEVER STORED. There is no state column (migration 080): a stored
// state is a second copy of what these three columns already say, and the only
// thing a second copy can do is disagree. Every surface that reports a state —
// the buyer's list, the Organization's guest list, the export — comes through
// here, so "assigned" cannot mean one thing on one page and another elsewhere.
//
// THE ORDER OF THE TESTS IS THE MODEL. Acceptance is checked first because it is
// the strongest fact: a Ticket that has been accepted has necessarily been
// assigned, and reporting the weaker fact would lose the Holder.
func AssignmentState(holderEmail string, assignedAt, acceptedAt *time.Time) TicketAssignmentState {
	if acceptedAt != nil {
		return TicketAccepted
	}
	if holderEmail != "" && assignedAt != nil {
		return TicketAssigned
	}
	return TicketUnassigned
}

// The Sales Channels a Ticket Sale can have been recorded on (migration 010).
//
// Named here rather than compared as bare strings because assignment is the
// first feature to REFUSE one of them, and a channel test written as a literal
// in three places is three chances to spell it differently.
const (
	SalesChannelOnline   = "online"
	SalesChannelInPerson = "in_person"
	SalesChannelImport   = "import"
)

// AssignmentRefusal reports whether a Ticket may be assigned at all, and why not.
//
// A SEPARATE ENUM FROM AnswerRefusal even though two of its members read alike,
// because the two windows are separate rules that happen to agree today. The
// answer window has no opinion about the Sales Channel — an `in_person` Ticket's
// Answers are perfectly writable by Event Staff — and folding assignment into
// AnswerWindow would either close that door or smuggle a channel test into a
// function three other callers depend on.
type AssignmentRefusal int

const (
	// AssignmentWindowOpen: this Ticket may be assigned and reassigned.
	AssignmentWindowOpen AssignmentRefusal = iota
	// AssignmentRefusedChannel: the Ticket Sale was recorded at the door.
	//
	// `in_person` HAS NO BUYER SURFACE, and that is the whole reason. Assignment
	// happens after purchase, from the Confirmation Link page or the Customer
	// Area, and a door sale gives the buyer neither: the POS is unbuilt and
	// `in_person` exists today only as a Sales Channel value (see
	// integration/tickets_test.go). Refusing here is therefore stating a fact
	// about the sale rather than withholding a feature.
	AssignmentRefusedChannel
	// AssignmentRefusedSaleReversed: the Ticket's Ticket Sale has been reversed.
	// Nobody is coming on a ticket that was refunded, so there is nobody to hand
	// it to.
	AssignmentRefusedSaleReversed
	// AssignmentRefusedEventStarted: the Event has started.
	//
	// The window closes at the doors, exactly as the Answer window does and for
	// the same reason: assignment exists so that the right person is named
	// before the Organization has to act on it, and the last moment that is any
	// use is the moment the doors open. It is also when #322's retention purge
	// takes an unaccepted address, so an assignment made after it would be
	// naming somebody the platform is in the act of forgetting.
	AssignmentRefusedEventStarted
)

// AssignmentWindow is the single place the assignment window is decided, for
// every route into a Ticket Assignment.
//
// THE CHANNEL IS TESTED FIRST because it is the most fundamental fact: an
// `in_person` sale never had a buyer surface to assign from, whether or not its
// Event has started, and telling that buyer "the event has started" would send
// them looking for a deadline they never had. Same reasoning as AnswerWindow's
// reversal-before-clock ordering, one rung further down.
//
// eventStartsAt is nil for an Event that has not said when it starts. That Event
// has not started — reading nil as "started" would freeze assignment on every
// unscheduled Event, which is the wrong way to be wrong.
func AssignmentWindow(
	channel, saleStatus string,
	eventStartsAt *time.Time,
	now time.Time,
) AssignmentRefusal {
	if channel != SalesChannelOnline && channel != SalesChannelImport {
		return AssignmentRefusedChannel
	}
	if saleStatus != TicketSaleStatusActive {
		return AssignmentRefusedSaleReversed
	}
	if eventStartsAt != nil && !now.Before(*eventStartsAt) {
		return AssignmentRefusedEventStarted
	}
	return AssignmentWindowOpen
}

// ParseHolderEmail turns what the buyer typed into the address that will be
// stored, or reports that it is not one.
//
// IT NORMALISES THROUGH platform.NormalizeEmail, the single point at which an
// address is trimmed and lowercased on this platform. That matters more here
// than anywhere: #325 mints or matches a Customer from a click at this address,
// and an address stored as the buyer capitalised it would fail to match the
// Customer record the same person already has — quietly turning one person into
// two.
//
// THE SHAPE CHECK IS mail.ParseAddress, as at every other door an address comes
// through (checkout, Sale Import, passcode sign-in). It is a weak check by
// design: an address is only really validated by mail arriving at it, and this
// one belongs to somebody who is not here to confirm anything. What it catches
// is the typo with no `@` in it, which is worth catching because until #325
// mails the address NOBODY WILL DISCOVER IT IS WRONG — the buyer sees their own
// typo echoed back and believes the job done.
//
// It refuses the display-name form (`Ana <ana@example.com>`) that
// mail.ParseAddress otherwise accepts, because what is stored must be an address
// and nothing else: the extra text would travel into a mail header in #325 and
// into the Organization's export, and neither is a place for a string the buyer
// pasted out of their own address book.
func ParseHolderEmail(raw string) (string, bool) {
	email := platform.NormalizeEmail(raw)
	if email == "" || len(email) > MaxHolderEmailLength {
		return "", false
	}
	parsed, err := mail.ParseAddress(email)
	if err != nil {
		return "", false
	}
	// ParseAddress accepts `Name <addr>` and reports the bare address; the two
	// agree only when the caller typed the bare address to begin with.
	if platform.NormalizeEmail(parsed.Address) != email {
		return "", false
	}
	return email, true
}
